Tests/NIOHTTP2Tests/SimpleClientServerTests.swift

```diff
@@ -1782,4 +1785,107 @@ class SimpleClientServerTests: XCTestCase {
 
         try self.clientChannel.assertReceivedFrame().assertGoAwayFrame(lastStreamID: .maxID, errorCode: UInt32(HTTP2ErrorCode.protocolError.networkCode), opaqueData: nil)
     }
+
+    func testNoStreamWindowUpdateOnEndStreamFrameFromServer() throws {
+        let maxFrameSize = (1 << 24) - 1  // Max value of SETTINGS_MAX_FRAME_SIZE is 2^24-1.
+        let clientSettings = nioDefaultSettings + [HTTP2Setting(parameter: .maxFrameSize, value: maxFrameSize)]
+
+        // Begin by getting the connection up.
+        // We add the stream multiplexer here because it manages automatic flow control, and because we want the server
+        // to correctly confirm we don't receive a frame for the child stream.
+        let childHandler = FrameRecorderHandler()
+        try self.basicHTTP2Connection(clientSettings: clientSettings) { channel, streamID in
+            // Here we send a large response: 65535 bytes in size.
+            let responseHeaders = HPACKHeaders([(":status", "200"), ("content-length", "65535")])
+
+            var responseBody = self.clientChannel.allocator.buffer(capacity: 65535)
+            responseBody.writeBytes(Array(repeating: UInt8(0x04), count: 65535))
+
+            let respFrame = HTTP2Frame(streamID: streamID, payload: .headers(.init(headers: responseHeaders)))
+            channel.write(respFrame, promise: nil)
+
+            // Now prepare the large body.
+            let respBodyFrame = HTTP2Frame(streamID: streamID, payload: .data(.init(data: .byteBuffer(responseBody), endStream: true)))
+            channel.writeAndFlush(respBodyFrame, promise: nil)
+
+            return channel.pipeline.addHandler(childHandler)
+        }
+
+        // Now we're going to send a small request.
+        let headers = HPACKHeaders([(":path", "/"), (":method", "POST"), (":scheme", "https"), (":authority", "localhost")])
+
+        // We're going to open a stream and queue up the frames for that stream.
+        let handler = try self.clientChannel.pipeline.context(handlerType: HTTP2StreamMultiplexer.self).wait().handler as! HTTP2StreamMultiplexer
+        var reqFrame: HTTP2Frame? = nil
+
+        handler.createStreamChannel(promise: nil) { channel, streamID in
+            // We need END_STREAM set here, because that will force the stream to be closed on the response.
+            reqFrame = HTTP2Frame(streamID: streamID, payload: .headers(.init(headers: headers, endStream: true)))
+            channel.writeAndFlush(reqFrame, promise: nil)
+            return channel.eventLoop.makeSucceededFuture(())
```

Comment:
Is there any reason we can't just return `channel.writeAndFlush(reqFrame)` here?

Reply:
Yes, that future is not guaranteed to immediately succeed if we're not careful, which can lead to a deadlock in this test. We don't want to wait for the write to be succeeded.